cli/internal/daemon/daemon.go

```diff
@@ -0,0 +1,324 @@
+}
+
+func addDaemonSubcommands(cmd *cobra.Command, config *config.Config, output cli.Ui) {
+	addStatusCmd(cmd, config, output)
+	addStartCmd(cmd, config, output)
+	addStopCmd(cmd, config, output)
+	addRestartCmd(cmd, config, output)
+}
+
+var errInactivityTimeout = errors.New("turbod shut down from inactivity")
+
+// debounceServers attempts to ensure that only one daemon is running from the given pid file path
+// at a time.
+func (d *daemon) debounceServers(pidPath fs.AbsolutePath) (lockfile.Lockfile, error) {
+	lockFile, err := lockfile.New(pidPath.ToString())
+	if err != nil {
+		return "", err
+	}
+	if err := lockFile.TryLock(); err != nil {
+		return "", err
+	}
+	return lockFile, nil
+
+}
+
+type rpcServer interface {
+	Register(grpcServer server.GRPCServer)
+}
+
+func (d *daemon) runTurboServer(parentContext context.Context, rpcServer rpcServer, signalWatcher *signals.Watcher) error {
+	ctx, cancel := context.WithCancel(parentContext)
+	defer cancel()
+	pidPath := getPidFile(d.repoRoot)
+	if err := pidPath.EnsureDir(); err != nil {
+		return err
+	}
+	lockFile, err := d.debounceServers(pidPath)
+	if err != nil {
+		return err
+	}
+	signalWatcher.AddOnClose(func() {
+		d.unlockPid(lockFile)
+	})
+	panicHandler := func(thePanic interface{}) error {
+		cancel()
+		d.logger.Error(fmt.Sprintf("Caught panic %v", thePanic))
+		return status.Error(codes.Internal, "server panicked")
+	}
+	defer d.unlockPid(lockFile)
+	// If we have the lock, assume that we are the owners of the socket file,
+	// whether it already exists or not. That means we are free to remove it.
+	sockPath := getUnixSocket(d.repoRoot)
+	if err := sockPath.Remove(); err != nil && !errors.Is(err, os.ErrNotExist) {
+		return err
+	}
+	d.logger.Debug(fmt.Sprintf("Using socket path %v (%v)\n", sockPath, len(sockPath)))
+	lis, err := net.Listen("unix", sockPath.ToString())
+	if err != nil {
+		return err
+	}
+	// We don't need to explicitly close 'lis', the grpc server will handle that
+	s := grpc.NewServer(
+		grpc.ChainUnaryInterceptor(
+			d.onRequest,
+			grpc_recovery.UnaryServerInterceptor(grpc_recovery.WithRecoveryHandler(panicHandler)),
+		),
+	)
+	signalWatcher.AddOnClose(s.GracefulStop)
+	go d.timeoutLoop(ctx)
+
+	rpcServer.Register(s)
+	errCh := make(chan error)
+	go func(errCh chan<- error) {
+		if err := s.Serve(lis); err != nil {
+			errCh <- err
+		}
+		close(errCh)
+	}(errCh)
+
+	var exitErr error
+	select {
+	case err, ok := <-errCh:
+		if ok {
+			exitErr = err
+		}
+	case <-d.timedOutCh:
+		exitErr = errInactivityTimeout
+		s.GracefulStop()
+	case <-ctx.Done():
+		s.GracefulStop()
+	}
+	// Wait for the server to exit, if it hasn't already.
+	// When it does, this channel will close. We don't
+	// care about the error in this scenario because we've
+	// either requested a close via cancelling the context
+	// or an inactivity timeout.
+	for range errCh {
+	}
+	return exitErr
+}
+
+func (d *daemon) unlockPid(lockFile lockfile.Lockfile) {
+	d.cleanup.Do(func() {
```

Comment:
```suggestion
	// Runs at most once.
	d.cleanup.Do(func() {
```

I assume I'm missing something because there's a lot of ceremony here to make sure that both of those unlock invocation locations play nicely.